pkg/auth/dependency/provider/password/authdata.go

```diff
@@ -55,14 +57,21 @@ var (
 // example 3: authRecordKeys = [["username", "email"], ["nickname"]]
 // - if authData is { "username": "john.doe", "nickname": "john.doe" },
 // output is [{ "nickname": "john.doe" }}]
+//
+// example 4: authRecordKeys = [["username"], ["email"]]
+// - if authData is { "username": "john.doe", "emamil": "" },
+// output is [{ "username": "john.doe" }}]
 func toValidAuthDataList(authRecordKeys [][]string, authData map[string]interface{}) []map[string]interface{} {
 	outputs := make([]map[string]interface{}, 0)
 
 	for _, ks := range authRecordKeys {
 		m := make(map[string]interface{})
 		for _, k := range ks {
 			for dk := range authData {
-				if k == dk && authData[dk] != nil {
+				if k == dk &&
+					authData[dk] != nil &&
+					// for non-nil interface but underlying value is a zero value, it won't be considered as a valid authData
+					!reflect.DeepEqual(authData[dk], reflect.Zero(reflect.TypeOf(authData[dk])).Interface()) {
```

Comment:
In the original design, we haven't restrict the type the auth data. But there are some problems.
1. For some data type like boolean, it will be rejected here. But it should be valid.
2. Since there is no restriction, user may pass some strange data here like json object. It is not the use case we want to support.
3. We don't have user defined validation to auth data, there will have problem if user pass some non-string data to email when sending welcome email.

We had a short chat on this, and we think string can fulfill most of the use case already. In order to keep it simple, maybe we can restrict user using string in auth data. Thought?

Reply:
I think it is okay if we don't expect `authData` in other data type. Do you want me to replace all `authData map[string]interface{}` to `authData map[string]string` in current code base? or just do type check in this function?